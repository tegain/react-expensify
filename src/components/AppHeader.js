import React from 'react'
import { connect } from 'react-redux'
import { NavLink } from 'react-router-dom'
import { startLogout } from "../actions/authActions";

export const AppHeader = ({ startLogout }) => (
	<header>
		<h1>Expensify</h1>
		<ul>
			<li><NavLink to="/dashboard" activeClassName="is-active">Dashboard</NavLink></li>
			<li><NavLink to="/create" activeClassName="is-active">Add new expense</NavLink></li>
			<li><button onClick={startLogout}>Logout</button></li>
		</ul>
	</header>
)

const mapDispatchToProps = (dispatch) => ({
	startLogout: () => dispatch(startLogout())
})


export default connect(undefined, mapDispatchToProps)(AppHeader)