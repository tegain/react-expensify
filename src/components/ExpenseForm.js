import React from 'react'
import moment from 'moment'
import 'react-dates/initialize'
import { SingleDatePicker } from 'react-dates'

export default class ExpenseForm extends React.Component {

	constructor (props) {
		super (props)

		this.state = {
			description: props.expense ? props.expense.description : '',
			note: props.expense ? props.expense.note : '',
			amount: props.expense ? (props.expense.amount / 100).toString() :'',
			createdAt: props.expense ? moment(props.expense.createdAt) : moment(),
			calendarFocused: false,
			error: undefined
		}
	}

	onDescriptionChange = (e) =>Â {
		const description = e.target.value
		this.setState(() => ({ description }))
	}

	onNoteChange = (e) => {
		const note = e.target.value
		this.setState(() => ({ note }))
	}

	onAmountChange = (e) => {
		const amount = e.target.value

		/**
		 * Starts (^) with digit no matter how long (but at least one, to infinity): \d{1,}
		 * Then takes optional group ()?, allowing 0 or 1 point, followed by 0 to 2 digits: (\.\d{0,2})?
		 * then finishes the expression: $
		 * @type {RegExp}
		 */
		const regex = /^\d{1,}(\.\d{0,2})?$/
		if (!amount || amount.match(regex)) { // '!amount' allows empty value, so the user can delete the amount
			this.setState(() => ({ amount }))
		}
	}

	onDateChange = (createdAt) => {
		if (createdAt) {
			this.setState(() => ({ createdAt }))
		}
	}

	onFocusChange = ({ focused }) => {
		console.log({focused})
		this.setState(() => ({ calendarFocused: focused }))
	}

	onSubmit = (e) => {
		e.preventDefault()

		if (!this.state.description || !this.state.amount) {
			// Set error
			this.setState(() => ({ error: 'Please provide description and amount' }))
		} else {
			// Clear error
			this.setState(() => ({ error: undefined }))

			/**
			 * Dispatch to prop method, so the Redux action actually happens on the AddExpensePage,
			 * and not on this form (which is supposed to be reusable for AddExpense and EditExpense)
			 */
			this.props.onSubmit({
				description: this.state.description,
				amount: parseFloat(this.state.amount) * 100,
				createdAt: this.state.createdAt.valueOf(), // Get timestamps from moment.js method
				note: this.state.note
			})
		}
	}

	render () {
		return (
			<div className="input-group">
				{this.state.error && <p>{this.state.error}</p>}

				<form onSubmit={this.onSubmit} className="expenses-form">
					<div className="input-group__item">
						<input
							type="text"
							className="text-input"
							placeholder="Description"
							autoFocus
							value={this.state.description}
							onChange={this.onDescriptionChange}
						/>
					</div>

					<div className="input-group__item">
						<input
							type="text"
							className="text-input"
							placeholder="Amount"
							value={this.state.amount}
							onChange={this.onAmountChange}
						/>
					</div>

					<div className="input-group__item">
						<SingleDatePicker
							date={this.state.createdAt}
							onDateChange={this.onDateChange}
							focused={this.state.calendarFocused}
							onFocusChange={this.onFocusChange}
							numberOfMonths={1}
							isOutsideRange={(day) => false} // Enables to select dates in the past
						/>
					</div>

					<div>
						<textarea
							placeholder="Add a note for your expense (optional)"
							value={this.state.note}
							onChange={this.onNoteChange}
						>
						</textarea>
					</div>

					<div>
						<button
							type="submit"
							className="button button--blue"
						>
							Save Expense
						</button>
					</div>
				</form>
			</div>
		)
	}

}