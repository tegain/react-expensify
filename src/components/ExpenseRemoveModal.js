import React from 'react'
import { connect } from 'react-redux'
import Modal from 'react-modal'
import {startRemoveExpense} from "../actions/expensesActions";

export class ExpenseRemoveModal extends React.Component {
	onRemoveExpense = () => {
		this.props.startRemoveExpense({ id: this.props.removeExpenseId })
		this.props.history.push('/')
	}

	render () {
		return (
			<Modal
				appElement={document.getElementById('root')}
				isOpen={!!this.props.removeExpenseId}
				onRequestClose={this.props.onClearRemoveExpense}
				contentLabel="Remove expense"
				closeTimeoutMS={200}
				className="modal"
			>
				<h3 className="ReactModal__title">Remove an expense</h3>
				<div className="ReactModal__inner">
					<p>Remove expense <strong>{this.props.removeExpenseDesc}</strong> ?</p>
					<div className="ReactModal__actions">
						<button
							className="button button--blue"
							onClick={this.onRemoveExpense}
						>
							Remove
						</button>

						<button
							className="button button--transparent"
							onClick={this.props.onClearRemoveExpense}
						>
							Cancel
						</button>
					</div>
				</div>
			</Modal>
		)
	}
}

const mapDispatchToProps = (dispatch) => ({
	startRemoveExpense: (data) => dispatch(startRemoveExpense(data))
})

export default connect(undefined, mapDispatchToProps)(ExpenseRemoveModal)