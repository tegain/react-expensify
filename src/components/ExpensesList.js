import React from 'react'
import { connect } from 'react-redux'
import ExpensesListItem from './ExpensesListItem'
import selectExpenses from '../selectors/expensesSelectors'

// Takes props from the Connected Component
export const ExpensesList = (props) => (
	<div className="expenses">
		<div className="expenses-header">
			<span className="expenses-header__expense">Expense</span>
			<span className="expenses-header__amount">Amount</span>
		</div>

		<div className="expenses-list">
			{
				props.expenses.length === 0 ? (
					<p className="expenses-list__empty">No expenses</p>
				) : (
					props.expenses.map((expense) => (
						<ExpensesListItem key={expense.id} {...expense} /> // Deconstruct expenses properties
					))
				)
			}
		</div>
	</div>
)

/**
 * This is a function that wraps your component in another component,
 * which subscribes to changes in the Redux store and renders itself and
 * consequently its descendants whenever an update occurs.
 *
 * connect(what_we_want_to_connect)(which_component_we_want_to_connect)
 */
// const ConnectedExpensesList = connect((state) => {
// 	return {
// 		expenses: state.expenses // We only want to connect state.expenses (not all the state)
// 	}
// })(ExpensesList)
//
// export default ConnectedExpensesList

// REFACTOR -->

const mapStateToProps = (state) => {
	return {
		expenses: selectExpenses(state.expenses, state.filters) // We only want to connect the filtered state.expenses (not all the state)
	}
}

export default connect(mapStateToProps)(ExpensesList)

