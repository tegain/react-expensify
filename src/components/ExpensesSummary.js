import React from 'react'
import { Link } from 'react-router-dom'
import { connect } from 'react-redux'
import numeral from 'numeral'
import '../locales/fr'
import getVisibleExpenses from '../selectors/expensesSelectors'
import getExpensestotal from '../selectors/expensesTotalSelectors'

export const ExpensesSummary = ({ expensesCount, expensesTotal }) => {
	const expenseWord = expensesCount === 1 ? 'expense' : 'expenses'
	const formattedExpensesTotal = numeral(expensesTotal / 100).format("0,0.[00] $")

	return (
		<div className="content-header">
			<div className="content-header__container container">
				<h2 className="content-header__heading">
					Viewing <strong>{expensesCount} {expenseWord}</strong> totalling <strong>{formattedExpensesTotal}</strong>
				</h2>

				<div className="content-header__actions">
					<Link
						to="/create"
						className="content-header__btn button button--blue"
					>
						Add expense
					</Link>
				</div>
			</div>
		</div>
	)
}

const mapStateToProps = (state) => {
	const visibleExpenses = getVisibleExpenses(state.expenses, state.filters)

	return {
		expensesCount: visibleExpenses.length,
		expensesTotal: getExpensestotal(visibleExpenses)
	}
}

export default connect(mapStateToProps)(ExpensesSummary)