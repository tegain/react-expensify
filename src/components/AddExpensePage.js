import React from 'react'
import { connect } from 'react-redux'
import ExpenseForm from './ExpenseForm'
import { startAddExpense } from '../actions/expensesActions'

export class AddExpensePage extends React.Component {
	onSubmit = (expense) => {
		this.props.startAddExpense(expense) // Dispatch startAddExpense to the store
		this.props.history.push('/') // Redirects to the dashboard page
	}

	render () {
		return (
			<div>
				<div className="content-header">
					<div className="content-header__container container">
						<h2 className="content-header__heading">Add new expense</h2>
					</div>
				</div>

				<div className="content-body">
					<div className="content-body__container container">
						<ExpenseForm
							onSubmit={this.onSubmit}
						/>
					</div>
				</div>
			</div>
		)
	}
}

// Use mapDispatchToProps to make the dispatch action independant from the comonent itself
// It makes the component easier to test, because there is only one linked method to test
const mapDispatchToProps = (dispatch) => ({
	startAddExpense: (expense) =>Â dispatch(startAddExpense(expense))
})

export default connect(undefined, mapDispatchToProps)(AddExpensePage)