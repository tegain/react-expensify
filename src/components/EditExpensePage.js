import React from 'react'
import { connect } from 'react-redux'
import {startEditExpense, startRemoveExpense} from '../actions/expensesActions'
import ExpenseForm from './ExpenseForm'

export class EditExpensePage extends React.Component {
	onSubmit = (expense) => {
		this.props.startEditExpense(this.props.expense.id, expense)
		this.props.history.push('/')
	}

	onRemove = () => {
		this.props.startRemoveExpense({ id: this.props.expense.id })
		this.props.history.push('/')
	}

	render () {
		return (
			<div>
				<div className="content-header">
					<div className="content-header__container container">
						<h2 className="content-header__heading">Edit expense</h2>
					</div>
				</div>

				<div className="content-body">
					<div className="content-body__container container">
						<ExpenseForm
							expense={this.props.expense}
							onSubmit={this.onSubmit}
						/>
						<button
							onClick={this.onRemove}
							className="button button--ghost"
						>
							Remove expense
						</button>
					</div>
				</div>
			</div>
		)
	}
}

const mapStateToProps = (state, props) => ({
	expense: state.expenses.find((expense) => expense.id === props.match.params.id)
})

const mapDispatchToProps = (dispatch, props) => ({
	startEditExpense: (id, expense) => dispatch(startEditExpense(id, expense)),
	startRemoveExpense: (data) => dispatch(startRemoveExpense(data))
})

export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage)