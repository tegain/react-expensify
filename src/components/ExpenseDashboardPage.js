import React from 'react'
import ExpensesSummary from './ExpensesSummary'
import ExpensesList from './ExpensesList'
import ExpensesListFilters from './ExpensesListFilters'

const ExpenseDashboardPage = () => (
	<div>
		<ExpensesSummary />
		<div className="container">
			<ExpensesListFilters />
			<ExpensesList />
		</div>
	</div>
)

export default ExpenseDashboardPage