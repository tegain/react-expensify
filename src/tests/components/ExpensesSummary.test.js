import React from 'react'
import { shallow } from 'enzyme'
import { ExpensesSummary } from '../../components/ExpensesSummary'

let wrapper, total, count

beforeEach(() => {
	total = 1750
	count = 1
	wrapper = shallow(<ExpensesSummary expensesTotal={total} expensesCount={count} />)
})

test('should render ExpensesSummary correctly with one expense', () => {
	expect(wrapper).toMatchSnapshot()
})

test('should render ExpensesSummary correctly with multiple expenses', () => {
	wrapper.setProps({ 'expensesCount': 3 })
	expect(wrapper).toMatchSnapshot()
})