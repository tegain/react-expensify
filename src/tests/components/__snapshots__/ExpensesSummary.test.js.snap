// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render ExpensesSummary correctly with multiple expenses 1`] = `
<div
  className="content-header"
>
  <div
    className="content-header__container container"
  >
    <h2
      className="content-header__heading"
    >
      Viewing 
      <strong>
        3
         
        expenses
      </strong>
       totalling 
      <strong>
        17,5 €
      </strong>
    </h2>
    <div
      className="content-header__actions"
    >
      <Link
        className="content-header__btn button button--blue"
        replace={false}
        to="/create"
      >
        Add expense
      </Link>
    </div>
  </div>
</div>
`;

exports[`should render ExpensesSummary correctly with one expense 1`] = `
<div
  className="content-header"
>
  <div
    className="content-header__container container"
  >
    <h2
      className="content-header__heading"
    >
      Viewing 
      <strong>
        1
         
        expense
      </strong>
       totalling 
      <strong>
        17,5 €
      </strong>
    </h2>
    <div
      className="content-header__actions"
    >
      <Link
        className="content-header__btn button button--blue"
        replace={false}
        to="/create"
      >
        Add expense
      </Link>
    </div>
  </div>
</div>
`;
